lms/djangoapps/course_api/blocks/forms.py

```diff
@@ -106,30 +107,70 @@ def clean(self):
         cleaned_data['user'] = self._clean_requested_user(cleaned_data, usage_key.course_key)
         return cleaned_data
 
+    def clean_username(self):
+        """
+        Return cleaned username.
+
+        Overrides the default behaviour that maps an empty string to None. This
+        allows us to differentiate between no username being provided (None) vs
+        an empty username being provided ('').
+        """
+        if 'username' in self.data:
+            return self.cleaned_data['username']
+        else:
+            return None
+
     def _clean_requested_user(self, cleaned_data, course_key):
         """
         Validates and returns the requested_user, while checking permissions.
         """
         requesting_user = self.initial['requesting_user']
         requested_username = cleaned_data.get('username', None)
+        all_blocks = cleaned_data.get('all_blocks', False)
+
+        if requested_username is not None and all_blocks:
+            raise ValidationError('The all_blocks and username options are mutually exclusive.')
+
+        if requested_username is None and not all_blocks:
+            raise ValidationError({'username': [u"This field is required unless all_blocks is requested."]})
 
-        if not requested_username:
-            return self._verify_no_user(requesting_user, cleaned_data, course_key)
+        if requesting_user.is_anonymous:
+            return self._verify_anonymous_user(requested_username, course_key, all_blocks)
+
+        if requested_username is None:
+            return self._verify_all_bocks(requesting_user, course_key)
         elif requesting_user.username.lower() == requested_username.lower():
             return self._verify_requesting_user(requesting_user, course_key)
         else:
             return self._verify_other_user(requesting_user, requested_username, course_key)
 
     @staticmethod
-    def _verify_no_user(requesting_user, cleaned_data, course_key):
+    def _verify_anonymous_user(username, course_key, all_blocks):
         """
-        Verifies form for when no username is specified, including permissions.
+        Verifies form for when the requesting user is anonymous.
         """
-        # Verify that access to all blocks is requested
-        # (and not unintentionally requested).
-        if not cleaned_data.get('all_blocks', None):
-            raise ValidationError({'username': ['This field is required unless all_blocks is requested.']})
+        if all_blocks:
+            raise PermissionDenied(
+                u"Anonymous users do not have permission to access all blocks in '{course_key}'."
+                .format(course_key=six.text_type(course_key))
+            )
+
+        if not username == '':
```

Comment:
This is "un-pythonic" (non-idiomatic Python)

```suggestion
        if username != '':
```